### ISSUE TITLE
Upward swipes sometimes don't register

### ISSUE THREAD

**Issue by StevenH237** [dessalines/thumb-key#937]
### Steps to reproduce

1. Use a layout that has up-swipes from the top row, such as programmer
2. Set the swipe length to a number like 70
3. Swipe upwards on a top-row key such as r, starting your swipe somewhere near the top of the key
4. Swipe up-left on a top-row key such as r, again starting your swipe somewhere near the top of the key

### Expected behavior

The upward (3) and up-left (4) gestures should be input, such as `^` or `"`

### Actual behavior

The tap (3) or left (4) gestures are input, such as `r` or `<`.

### version of the program

3.2.1

### Android version

Android 13

### Device

Google Pixel 6a

### Other details

Phone is rooted and using LineageOS 20-20231104-NIGHTLY-bluejay (I should really update that...)

It feels to me like the phone is treating the swipe as never going higher than the top of the keyboard. Notably, a swipe upwards from the *bottom* of the key will work, same as a swipe upwards from the bottom of the h key will still produce a q even if you don't swipe into the r key, but it is noticeably harder to get an up-swipe from the top row than from another row.

### Acknowledgements

- [X] I have written a short but informative title.
- [X] I have updated the app to **[the latest version](https://github.com/user/repo/releases/latest)**.
- [X] I have searched the existing issues and this is a new one, **NOT** a  duplicate or related to another open issue.
- [X] This is not a question or a discussion, in which case I should have  gone to [lemmy.ml/c/thumbkey](https://lemmy.ml/c/thumbkey)
- [X] This is a **single** bug report, in case of multiple bugs I will open  a separate issue for each one  (they can always link to each other if related)
- [ ] I have admitted that I am a clown by having checked this box, as I have not read these acknowledgements. 🤡
- [X] I have filled out all of the requested information in this form.

**Comment by glm4610** [comment]
This is something I mentioned in another issue, which happens because of how Android clamps the pointer motion data to within the IME window. Because of that, we can't properly track user gestures which wander upwards. I think Unexpected Keyboard has a good workaround for this. I'll take a look at it this week

**Comment by breversa** [comment]
I’m having a similar issue with the FR v2 layout:

In the **top-left square**, the **top-right symbol** (the french apostrophe "`'`") rarely registers properly : it’s the right symbol ("`^`") that is input, unless I very precisely swipe in a not-totally-but-almost vertical way (= in the 1 o’clock direction from the center of the square).

As far as I know, it doesn’t happen with other squares/positions.

**Comment by Noitarud** [comment]
rightward guestures on right keys (redo, cut and whatever … is) and upward guestures on top (settings) do not work.

perhaps abstractly related: noted the rightward stroke ")" is difficult to enter, only possible if I swipe from left side instead of centre.

To fix rightward, move a little from edge. Topward doesn't matter but adds confusion, place on bksp or fixed down-right #

**Comment by pxeemo** [comment]
If I'm not mistaken this is because gestures are not registered out of the keyboard area which works fine in Android 14+ based on my experiments
Also scrolling emojis upwards out of the keyboard's area doesn't work in Android 13 which was fine in Android 14+

**Comment by mpsijm** [comment]
You're correct in saying that the positions of the gestured are clipped to be within the keyboard area, I also experimented with this a while ago.

Unfortunately, on my Android 14 phone (Samsung Galaxy A33, latest update as of today), this is not yet fixed... Scrolling in the emoji list still stops when my finger moves outside the keyboard's area.

**Comment by Nixill** [comment]
Expanding the keyboard area would just mean wasted space between the keyboard and whatever you're typing into, as apps that expect your textbox to be the bottom of the screen (other than the keyboard) would just not display anything in that extra space.

A workaround I just thought of is that you could treat a swipe *to* the top of the keyboard as if it goes to some arbitrary position above the top of the keyboard. I don't believe this perceived jump/"hook" motion would adversely affect diagonal swipes because outside of circular or return gestures I believe only the start and end positions are really checked anyway?

**Comment by tuxayo** [comment]
Thanks all, knowing the cause of the issue. Now I can more reliably make letters by starting my swipe at the bottom of the key :D
It still a workaround but it's much more usable.

note: releasing still withing the key seems to work better
**It seems when releasing outside the keyboard, it forgets that inside the keyboard, it was a diagonal.** So maybe it could be made more reliable still without being able to track movement above the key.

If that can help: very large switch outside the keyboard are terrible. If not going far above the keyboard, it seem it can still process well the part that was on the key

**Comment by tuxayo** [comment]
> Expanding the keyboard area would just mean wasted space between the keyboard and whatever you're typing into

If there is an implementation of #44 , it would solve this without additional work if the completions or corrections are listed above the keyboard.

**Comment by alerque** [comment]
> It feels to me like the phone is treating the swipe as never going higher than the top of the keyboard.

I'm also on LineageOS (although somewhat newer than reported in this issue).

I can confirm that this is **definitely not** the case. I can swipe straight up or even to the left from the top row, then after I am way out of the top of the keyboard area dog-leg over to the right and get the `topright` action to trigger.

Is this perhaps fixed already? Does anybody still actually have this issue as reproducible?

**Comment by dessalines** [comment]
I just tried and this does seem to be fixed (probably was a jetpack compose bug). Can re-open if necessary.
